### PR TITLE
haproxy: 2.9.2 -> 2.9.3; Enable QUIC support

### PR DESCRIPTION
**Pull request by vifino**
## Description of changes
This enables switching SSL Libraries.
In addition, this switches to PCRE2 and Lua 5.4 from PCRE and Lua 5.3.
Should have no functional differences, apart from the added QUIC support which needs to be configured in HAProxy.

This also adds more test cases, specifically for SSL, mutual TLS and QUIC.

According to the recent [SSL Library Support Wiki Page](https://github.com/haproxy/wiki/wiki/SSL-Libraries-Support-Status), OpenSSL 3.X is not the best choice by far.

I would really like WolfSSL, as that seems to be the one with the best performance yet there are still many unimplemented features.
It requires a different build config, but that's the way wolfssl intends to be used.

However, in order to make this easier to merge without worrying about breakage, it still defaults to OpenSSL (3.X) 
A future PR might flip the switch, though.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vifino** [comment]
I saw #202922, but that is IMO not the way to go.
OpenSSL 3.X is **not recommended** for HAProxy.
As a direct quote from [SSL Library Support Wiki Page](https://github.com/haproxy/wiki/wiki/SSL-Libraries-Support-Status#openssl):

> At this point, **OpenSSL 3.x is basically only usable for personal sites**. Users seeking anything serious performance-wise will need to **stay on version 1.1.1 provided by their vendor** or to recompile one of the alternate libraries below.

**Comment by vifino** [comment]
Put the effort in to make it work with WolfSSL, which is the best option at the moment.
Should be less user-facing breakage than LibreSSL, actually!

**Comment by vifino** [comment]
Actually, I'll rework the WolfSSL derivation first.

File: pkgs/development/libraries/wolfssl/default.nix [hunk 1]
```diff
@@ -6,73 +6,102 @@
...
 
-  postPatch = ''
-    patchShebangs ./scripts
-    # ocsp tests require network access
-    sed -i -e '/ocsp\.test/d' -e '/ocsp-stapling\.test/d' scripts/include.am
-    # ensure test detects musl-based systems too
-    substituteInPlace scripts/ocsp-stapling2.test \
-      --replace '"linux-gnu"' '"linux-"'
-  '';
+  buildWolfSSL =
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/wolfssl/default.nix:18]: Why do we need this helper? We only have one version and people can use overrideAttrs.

File: pkgs/development/libraries/wolfssl/default.nix [hunk 2]
```diff
@@ -6,73 +6,102 @@
...
 }:
-
-stdenv.mkDerivation rec {
-  pname = "wolfssl";
-  version = "5.6.3";
-
-  src = fetchFromGitHub {
+let
+  defaultVersion = "5.6.3";
+  defaultSrc = fetchFromGitHub {
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/wolfssl/default.nix:11]: we should keep the default names version and src

File: pkgs/development/libraries/wolfssl/default.nix [hunk 3]
```diff
@@ -6,73 +6,102 @@
...
+      postPatch = ''
+        patchShebangs ./scripts
+        # ocsp tests require network access
+        sed -i -e '/ocsp\.test/d' -e '/ocsp-stapling\.test/d' scripts/include.am
+        # ensure test detects musl-based systems too
+        substituteInPlace scripts/ocsp-stapling2.test \
+          --replace '"linux-gnu"' '"linux-"'
+      '';
+
+      configureFlags = (map (o: "--enable-${o}") ([
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/wolfssl/default.nix:39]: I would rather remove this map to make things easier and simpler when consulting the upstream doc

File: pkgs/development/libraries/wolfssl/default.nix [hunk 4]
```diff
@@ -6,73 +6,102 @@
...
+        "reproducible-build"
+      ] ++ lib.optionals (defaultFlags) [
+        "${variant}"
+      ] ++ lib.optionals (defaultFlags && stdenv.hostPlatform.isx86) [
+        # Enable AVX/AVX2/AES-NI instructions, gated by runtime detection via CPUID.
+        "intelasm"
+        "aesni"
+      ] ++ lib.optionals (defaultFlags && stdenv.isAarch64 && stdenv.isDarwin) [
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/wolfssl/default.nix:47]: Wouldn't this apply to aarch64-linux, too?

File: pkgs/development/libraries/wolfssl/default.nix [hunk 5]
```diff
@@ -6,73 +6,102 @@
...
+        # ocsp tests require network access
+        sed -i -e '/ocsp\.test/d' -e '/ocsp-stapling\.test/d' scripts/include.am
+        # ensure test detects musl-based systems too
+        substituteInPlace scripts/ocsp-stapling2.test \
+          --replace '"linux-gnu"' '"linux-"'
+      '';
+
+      configureFlags = (map (o: "--enable-${o}") ([
+        "reproducible-build"
+      ] ++ lib.optionals (defaultFlags) [
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/wolfssl/default.nix:41]: It is probably easier to just override the entire configureFlags with overrideAttrs than having this defaulFlags option.

File: pkgs/development/libraries/wolfssl/default.nix [hunk 6]
```diff
@@ -6,73 +6,102 @@
...
 
-  meta = with lib; {
-    description = "A small, fast, portable implementation of TLS/SSL for embedded devices";
-    homepage = "https://www.wolfssl.com/";
-    changelog = "https://github.com/wolfSSL/wolfssl/releases/tag/v${version}-stable";
-    platforms = platforms.all;
-    license = licenses.gpl2Plus;
-    maintainers = with maintainers; [ fab ];
+  wolfssl = buildWolfSSL {
+    # Almost same as Debian but for now using "all" instead of "distro" to ensure options.h gets installed
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/wolfssl/default.nix:98]: I am not sure why this matters to us that it is the same as debian.

File: pkgs/development/libraries/wolfssl/default.nix [hunk 7]
```diff
@@ -104,4 +104,9 @@ in
       "tls13"
     ];
   };
+
+  wolfssl-haproxy = buildWolfSSL {
+    variant = "haproxy";
+    extraOptions = [ "quic" ];
+  };
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/wolfssl/default.nix:111]:
```suggestion
  wolfssl-haproxy = wolfssl.overrideAttrs (oldAttrs: {
    pname = "wolfssl-haproxy";

    configureFlags = oldAttrs.configureFlags ++ [
      "--enable-haproxy"
      "--enable-quic"
    ];
  });
```
with this simple change we can almost completely discard the previous commit.

File: pkgs/tools/networking/haproxy/default.nix [hunk 8]
```diff
@@ -1,22 +1,39 @@
...
-, lua5_3 ? null
-, pcre ? null
+, wolfssl-haproxy ? null
+, libressl ? null
+, quictls ? null
+, boringssl ? null
+, openssl ? null
+, lua5_4 ? null
+, pcre2 ? null
 , systemd ? null
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/haproxy/default.nix:18]: ``? null`` is a legacy pattern with major pitfalls. It is not required and only hides bugs, please remove it.

File: pkgs/tools/networking/haproxy/default.nix [hunk 9]
```diff
@@ -1,22 +1,39 @@
...
-assert useLua -> lua5_3 != null;
-assert usePcre -> pcre != null;
+assert useLua -> lua5_4 != null;
+assert usePcre -> pcre2 != null;
+assert lib.assertOneOf "sslLibrary" sslLibrary [ "wolfssl" "libressl" "quictls" "boringssl" "openssl" ];
+assert sslLibrary == "wolfssl" -> wolfssl-haproxy != null;
+assert sslLibrary == "libressl" -> libressl != null;
+assert sslLibrary == "quictls" -> quictls != null;
+assert sslLibrary == "boringssl" -> boringssl != null;
+assert sslLibrary == "openssl" -> openssl != null;
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/haproxy/default.nix:28]: We don#t need any of the null checks here.

File: pkgs/tools/networking/haproxy/default.nix [hunk 10]
```diff
@@ -1,22 +1,39 @@
...
 
-stdenv.mkDerivation (finalAttrs: {
+let
+  sslPkgs = {
+    inherit libressl quictls boringssl openssl;
+    wolfssl = wolfssl-haproxy;
+  };
+  sslPkg = sslPkgs.${sslLibrary};
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/haproxy/default.nix:35]: We should probably just override the packages and override openssl like we do for nginx than this other layer of abstraction.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/wolfssl/default.nix (hunk 4 above)
Comment (vifino) [reply pkgs/development/libraries/wolfssl/default.nix]: Yes, but only on some machines.
IIRC the aarch64-linux target is `armv8+a`, so the crypto extensions aren't there. As there is no detection, we can't assume they exist.

**Comment by vifino** [comment]
Hey @SuperSandro2000, thank you kindly for reviewing.

The wolfssl changes are from #262594, I just needed to base the HAProxy changes on top - but I implemented your feedback (or tried, at least).

As for the `sslPkgs` stuff, I am not sure how to do it better as I'd need to detect if it's OpenSSL, OpenSSL with QuicTLS-compatible API (QuicTLS, BoringSSL, LibreSSL) or WolfSSL.
Do you have a suggestion?

**Comment by vifino** [comment]
I added some tests.

WolfSSL presumably requires a new release to get functioning HTTPS Client Certificate Authentication and I couldn't figure out if there's just a small patch to fix it between the 800+ commits that happened since the last release, so I think using it as default is only wise once that happens.
Sad, the performance difference is quite drastic!

No QUIC tests yet, as I couldn't seem to get them working at all.
`curlHTTP3` with the `--http3-only` flag didn't seem to want to connect.
QUIC does function with a browser in all the supported implementations (apart from OpenSSL, which doesn't seem to have one).

HAProxy does not seem to support cert auth under QUIC at the moment, unsure what causes that. I opened an issue upstream: https://github.com/haproxy/haproxy/issues/2322

**Comment by wlallemand** [comment]
Hello,

Quite honestly migrating to any other SSL library for HAProxy will remove a lot of features, neither LibreSSL nor WolfSSL implement the ClientHello callback which allows to have a dual ECDSA/RSA certificate for a same domain  for example.
And a lot of reg-tests are not working.

With WolfSSL only 15/30 SSL tests are working when using the latest 2.9-dev and the latest wolfssl commit.
It is a totally different stack not based on OpenSSL but they developed an OpenSSL compatibility API that does not provide the whole Openssl features. We hope to have a better support for it, but it's still considered experimental for HAProxy.

If you don't care about the performances issues of OpenSSL/quicTLS 3.x (it could be enough for small setups):

* QUIC with a full-featured haproxy you should consider openssl with USE_QUIC_OPENSSL_COMPAT=1. This will allow you to use QUIC with a standard openssl library (without the quictls patchset). The only limitation is that it does not provide the 0-rtt feature for QUIC.
It was developed in version 2.9-dev but I backported it in 2.8 so it will be available in the next 2.8.4 version.

* If you want a full-featured HAProxy with a full-featured QUIC, quictls is the best choice.

For other libraries:

* Regarding boringSSL we don't really work on  it anymore, the API changes a lot and it's a rolling release cycle, we don't even build it on the per-push CI.

* There is also the choice of AWSLC, but it's not in a release version yet, I'll probably backport the build fixes in 2.8 for it. We recently fixes the issue of QUIC for it, but it lacks some ciphers for QUIC.

Regards,

**Comment by vifino** [comment]
Hey @wlallemand, thank you so much for chiming in!

My current patch changes our default to QuicTLS as it is OpenSSL 3.X right now.
Thank you for confirming that it does result in a full-featured HAProxy with QUIC!
Perhaps changing back to OpenSSL 3.X after the next 2.8.4 release is something to consider.

This patch also allows one to optionally switch to a different implementation, like WolfSSL, but it won't be the default for a long time. I abandoned that idea pretty "quicly". ;)

But I will remove BoringSSL as an option in this patch, good to know that it is not something you really work on anymore!

Thank you so much!

**Comment by wlallemand** [comment]
Small update, we released haproxy 2.8.4, you can now enable USE_QUIC_OPENSSL_COMPAT=1 to have QUIC with an non-patched version of openssl.

**Comment by vifino** [comment]
Update on this:

- Bumped to 2.9.0, it removes the need for some workarounds.
- Enabled `USE_QUIC_OPENSSL_COMPAT` for OpenSSL. All providers enable QUIC now.
- Default to OpenSSL again to reduce breakage of existing setups to zero.

**Comment by wlallemand** [comment]
@vifino I'm not aware of any issue of this kind with the openssl compat layer, it could be a bug or a limitation of the current achitecture. Did you had the same issue with 2.8.4?  Better open an issue on the haproxy github indeed! thanks.

**Review by SuperSandro2000 — APPROVED** [review]

**Review by h7x4 — APPROVED** [review]
This looks really good. Thank you!

**Pull request merged**